src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsBuffer.cpp

```diff
@@ -0,0 +1,237 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+
+#include <NovelRT/Experimental/Graphics/Vulkan/Graphics.Vulkan.h>
+#include <NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsBuffer.h>
+
+namespace NovelRT::Experimental::Graphics::Vulkan
+{
+    VulkanGraphicsDevice* VulkanGraphicsBuffer::GetDeviceInternal() const noexcept
+    {
+        // TODO: This feels like it might be a perf issue later.
+        return dynamic_cast<VulkanGraphicsDevice*>(GraphicsDeviceObject::GetDevice().get());
+    }
+
+    VulkanGraphicsBuffer::VulkanGraphicsBuffer(std::shared_ptr<VulkanGraphicsDevice> device,
+                                               GraphicsBufferKind kind,
+                                               GraphicsMemoryRegion<GraphicsMemoryBlock> blockRegion,
+                                               GraphicsResourceAccess cpuAccess,
+                                               VkBuffer buffer)
+        : GraphicsBuffer(std::move(device), kind, std::move(blockRegion), cpuAccess), _vulkanBuffer(buffer)
+    {
+        VkResult bindResult = vkBindBufferMemory(GetAllocator()->GetDevice()->GetVulkanDevice(), _vulkanBuffer,
+                                                 GetBlock()->GetVulkanDeviceMemory(), GetBlockRegion().GetOffset());
+
+        if (bindResult != VK_SUCCESS)
+        {
+            throw Exceptions::InitialisationFailureException("Failed to bind the VkBuffer!", bindResult);
+        }
+    }
+
+    void* VulkanGraphicsBuffer::MapUntyped()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        void* pDestination = nullptr;
+
+        VkResult mapMemoryResult =
+            vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (mapMemoryResult != VK_SUCCESS)
+        {
+            // TODO: Make a real exception.
+            throw std::runtime_error("Failed to map Vulkan memory to the CPU!");
```

Comment:
Where's the "real" exception? Should it be Graphics / Vulkan specific?

Reply:
This is up for design discussion. I'll open a new ticket for this so we can design this later, but I think for this PR its out of scope. Thanks for reminding me to make an issue for it though.